model/task_lifecycle.go

```diff
@@ -2282,33 +2282,93 @@ func UpdateDisplayTaskForTask(t *task.Task) error {
 	if !t.IsPartOfDisplay() {
 		return errors.Errorf("task '%s' is not an execution task", t.Id)
 	}
-	dt, err := t.GetDisplayTask()
-	if err != nil {
-		return errors.Wrap(err, "getting display task for task")
+
+	// The display task status update can retry in case it temporarily
+	// conflicts with other execution tasks that are updating it at the same
+	// time (e.g. because end task is running for multiple execution tasks in
+	// parallel, so some see it as still running, but some see it as finished).
+	// While there's no exact number of times that this update can fail
+	// theoretically, the number of attempts is arbitrarily kept small, because
+	// the practical likelihood of an execution task racing more than once in a
+	// row is low.
+	const maxUpdateAttempts = 3
+	var (
+		originalDisplayTask *task.Task
+		updatedDisplayTask  *task.Task
+		err                 error
+	)
+	for i := 0; i < maxUpdateAttempts; i++ {
+		// Clear the cached display task, if any (e.g. due to a prior
+		// GetDisplayTask). The display task fetched here must always contain
+		// the latest display task data.
+		t.DisplayTask = nil
+
+		originalDisplayTask, err = t.GetDisplayTask()
+		if err != nil {
+			return errors.Wrap(err, "getting display task for task")
+		}
+		if originalDisplayTask == nil {
+			grip.Error(message.Fields{
+				"message":         "task may hold a display task that doesn't exist",
+				"task_id":         t.Id,
+				"display_task_id": t.DisplayTaskId,
+			})
+			return errors.Errorf("display task not found for task '%s'", t.Id)
+		}
+		if !originalDisplayTask.DisplayOnly {
+			return errors.Errorf("task '%s' is not a display task", originalDisplayTask.Id)
+		}
+
+		updatedDisplayTask, err = tryUpdateDisplayTaskAtomically(*originalDisplayTask)
+		if err == nil {
```

Comment:
If an execution task is operating on stale data I'd expect err to still be nil, so would this just be checking for a DB operation / read error?

Reply:
If it's operating on stale data, the UpdateOne should fail because [it will modify zero documents](https://github.com/mongodb/anser/blob/2a589a5299e680bcd5014909df16209967010365/db/wrapper.go#L206-L208).